nx/source/kernel/rwlock.c

```diff
@@ -3,33 +3,57 @@
 #include "kernel/rwlock.h"
 
 void rwlockInit(RwLock* r) {
-    rmutexInit(&r->r);
-    rmutexInit(&r->g);
-    r->b = 0;
+    mutexInit(&r->mutex);
+    condvarInit(&r->condvar_readers);
+    condvarInit(&r->condvar_writer);
+
+    r->readers = 0;
+    r->writer = false;
 }
 
 void rwlockReadLock(RwLock* r) {
-    rmutexLock(&r->r);
+    mutexLock(&r->mutex);
+
+    while (r->writer) {
+        condvarWait(&r->condvar_writer, &r->mutex);
+    }
 
-    if (r->b++ == 0)
-        rmutexLock(&r->g);
+    r->readers++;
 
-    rmutexUnlock(&r->r);
+    mutexUnlock(&r->mutex);
 }
 
 void rwlockReadUnlock(RwLock* r) {
-    rmutexLock(&r->r);
+    mutexLock(&r->mutex);
 
-    if (r->b-- == 1)
-        rmutexUnlock(&r->g);
+    if (r->readers-- == 1) {
+        condvarWakeAll(&r->condvar_readers);
+    }
 
-    rmutexUnlock(&r->r);
+    mutexUnlock(&r->mutex);
 }
 
 void rwlockWriteLock(RwLock* r) {
-    rmutexLock(&r->g);
+    mutexLock(&r->mutex);
+
+    while (r->writer) {
+        condvarWait(&r->condvar_writer, &r->mutex);
+    }
+
+    r->writer = true;
+
+    while (r->readers > 0) {
```

Comment:
`while (r->readers)`

Reply:
prefer not using boolean operators (!) on integer types